### PR TITLE
Feat: refactor resources/data to use the new serialization/deserializ…

### PR DESCRIPTION
**Pull request by TomerHeber**
…ation logic

### Solution

Most of this work was done in a pending PR https://github.com/env0/terraform-provider-env0/pull/393
However, in an effort to split the PR and unblock me from doing additional refactoring I've decided to create this PR.

Part of #376

### CODE REVIEW

File: env0/utils.go [hunk 1]
```diff
@@ -221,6 +221,94 @@ func writeResourceData(i interface{}, d *schema.ResourceData) error {
...
+		field := val.Index(i)
+		result = append(result, field.Interface())
+	}
+
+	return result
+}
+
+// Extracts values from a slice of interfaces, and writes it to resourcedata at name.
+func writeResourceDataSlice(i interface{}, name string, d *schema.ResourceData) error {
+	writeResourceDataSliceStruct := func(val reflect.Value, value map[string]interface{}) {
```
Comment (TomerHeber) [inline env0/utils.go:239]: Yeah this code is a bit complex.
Ping me if any clarifications are required.

File: env0/utils.go [hunk 2]
```diff
@@ -221,6 +221,94 @@ func writeResourceData(i interface{}, d *schema.ResourceData) error {
...
+	for i := 0; i < val.Len(); i++ {
+		field := val.Index(i)
+		result = append(result, field.Interface())
+	}
+
+	return result
+}
+
+// Extracts values from a slice of interfaces, and writes it to resourcedata at name.
+func writeResourceDataSlice(i interface{}, name string, d *schema.ResourceData) error {
```
Comment (yaronya) [inline env0/utils.go:238]: Why do we need a different func for slice? Can't we simply utilize the existing `writeResourceData`?

File: env0/utils.go [hunk 3]
```diff
@@ -221,6 +221,94 @@ func writeResourceData(i interface{}, d *schema.ResourceData) error {
...
+
+			if fieldType.Kind() == reflect.Ptr {
+				if field.IsNil() {
+					continue
+				}
+
+				field = field.Elem()
+			}
+
+			if d.Get(name+".*."+fieldNameSC) == nil {
```
[the remark this answers is not in the document]
Comment (yaronya) [reply env0/utils.go:254]: Same comment for the whole logic here actually, extracting it into smaller functions can be really helpful to figure what's going on here.

[the remark this answers is not in the document]
Comment (TomerHeber) [reply env0/utils.go:254]: Yup makes sense, let me see what I can do.

File: env0/utils.go (hunk 2 above)
Comment (TomerHeber) [reply env0/utils.go:238]: The end-goal is for writeResourceData to call this function.
I'll try and refactor it a little.

File: env0/utils.go [hunk 4]
```diff
@@ -117,29 +117,38 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
 	return nil
 }
 
+// Returns the field name or skip.
+func getFieldName(field reflect.StructField) (string, bool) {
```
Comment (TomerHeber) [inline env0/utils.go:121]: As discussed, I've tried moving common code to functions.
Added some more comments. It's still very complex, but I think it's much more readable now.

File: env0/utils.go [hunk 5]
```diff
@@ -117,29 +117,38 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
...
+		if skip {
+			continue
 		}
 
 		field := val.Field(i)
-
 		fieldType := field.Type()
 
-		if fieldName == "Id" {
+		if fieldName == "id" {
```
Comment (TomerHeber) [inline env0/utils.go:151]: uses the camalcase now....

File: env0/utils.go [hunk 6]
```diff
@@ -170,48 +179,20 @@ func writeResourceData(i interface{}, d *schema.ResourceData) error {
...
-				for i := 0; i < field.Len(); i++ {
-					str := field.Index(i).Interface().(string)
-					strs = append(strs, str)
-				}
-				if err := d.Set(fieldNameSC, strs); err != nil {
-					return err
-				}
-			default:
-				return fmt.Errorf("internal error - unhandled slice type %v", field.Type())
+			if err := writeResourceDataSlice(field.Interface(), fieldName, d); err != nil {
```
Comment (TomerHeber) [inline env0/utils.go:194]: This new method should cover all the above cases. There should be unit tests for all these cases.

File: env0/utils.go [hunk 7]
```diff
@@ -221,6 +202,95 @@ func writeResourceData(i interface{}, d *schema.ResourceData) error {
...
+		}
+
+		switch valType.Kind() {
+		case reflect.String:
+			values = append(values, val.String())
+		case reflect.Int:
+			values = append(values, val.Int())
+		case reflect.Bool:
+			values = append(values, val.Bool())
+		case reflect.Struct:
```
Comment (TomerHeber) [inline env0/utils.go:275]: This is a nontrivial case...
A slice of structs.

File: env0/utils.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomerHeber) [reply env0/utils.go]: Hopefully it's better now...

File: env0/utils.go (hunk 2 above)
Comment (TomerHeber) [reply env0/utils.go]: Applied some refactoring to make it more readable. Also incorporated the new function in writeResourceData

**Review by yaronya — APPROVED** [review]

**Pull request merged**